File: libgomp/testsuite/libgomp.c/examples-4/declare_target-4.c
/* { dg-do run } */
/* { dg-additional-options "-DTESTITERS=20" { target { ! run_expensive_tests } } } */

#include <stdlib.h>

#define EPS 0.00001
#define N 1000
#ifndef TESTITERS
#define TESTITERS N
#endif

#pragma omp declare target
float Q[N][N];
float Pfun (const int i, const int k)
{
  return Q[i][k] * Q[k][i];
}
#pragma omp end declare target

void init ()
{
  int i, j;
  for (i = 0; i < N; i++)
    for (j = 0; j < N; j++)
      Q[i][j] = 0.001 * i * j;
}

float accum_ref (int k)
{
  int i;
  float tmp = 0.0;

  for (i = 0; i < N; i++)
    tmp += Pfun (i, k);

  return tmp;
}

float accum (int k)
{
  int i;
  float tmp = 0.0;

  #pragma omp target map(tofrom:tmp)
    #pragma omp parallel for reduction(+:tmp)
      for (i = 0; i < N; i++)
	tmp += Pfun (i, k);

  return tmp;
}

void check (float a, float b)
{
  float err = (b == 0.0) ? a : (a - b) / b;
  if (((err > 0) ? err : -err) > EPS)
    abort ();
}

int main ()
{
  int i;

  init ();

  #pragma omp target update to(Q)

  for (i = 0; i < TESTITERS; i++)
    check (accum (i), accum_ref (i));

  return 0;
}

File: libgomp/testsuite/libgomp.fortran/examples-4/declare_target-4.f90
! { dg-do run }

module e_53_4_mod
  !$omp declare target (N, Q)
  integer, parameter :: N = 10
  real :: Q(N,N)
contains
  real function Pfun (i, k)
    !$omp declare target
    integer, intent(in) :: i, k
    Pfun = (Q(i,k) * Q(k,i))
  end function
end module

real function accum (k) result (tmp)
  use e_53_4_mod
  integer :: i, k
  tmp = 0.0e0
  !$omp target map(tmp)
    !$omp parallel do reduction(+:tmp)
    do i = 1, N
      tmp = tmp + Pfun (k, i)
    end do
  !$omp end target
end function

real function accum_ref (k) result (tmp)
  use e_53_4_mod
  integer :: i, k
  tmp = 0.0e0
  do i = 1, N
    tmp = tmp + Pfun (k, i)
  end do
end function

subroutine init ()
  use e_53_4_mod
  integer :: i, j
  do i = 1, N
    do j = 1, N
      Q(i,j) = 0.001 * i * j
    end do
  end do
end subroutine

subroutine check (a, b)
  real :: a, b, err
  real, parameter :: EPS = 0.00001
  if (b == 0.0) then
    err = a
  else if (a == 0.0) then
    err = b
  else
    err = (a - b) / b
  end if
  if (err > EPS .or. err < -EPS) stop 1
end subroutine

program e_53_4
  use e_53_4_mod
  integer :: i
  real :: accum, accum_ref
  call init ()
  !$omp target update to(Q)
  do i = 1, N
    call check (accum (i), accum_ref (i))
  end do
end program
